src/collapse/common.rs

```diff
@@ -0,0 +1,794 @@
+use std::borrow::Cow;
+use std::io;
+use std::mem;
+use std::sync::Arc;
+
+use chashmap::CHashMap;
+use crossbeam::channel;
+use fnv::FnvHashMap;
+use lazy_static::lazy_static;
+
+const CAPACITY_HASHMAP: usize = 512;
+pub(crate) const CAPACITY_READER: usize = 128 * 1024;
+pub(crate) const DEFAULT_NSTACKS_PER_JOB: usize = 100;
+const NBYTES_PER_STACK_GUESS: usize = 1024;
+const RUST_HASH_LENGTH: usize = 17;
+
+lazy_static! {
+    #[doc(hidden)]
+    pub static ref DEFAULT_NTHREADS: usize = num_cpus::get();
+}
+
+/// Private trait for internal library authors.
+///
+/// If you implement this trait, your type will implement the public-facing
+/// `Collapse` trait as well. Implementing this trait gives you parallelism
+/// for free as long as you adhere to the requirements described in the
+/// comments below.
+pub trait CollapsePrivate: Clone + Send + Sized + Sized {
+    // *********************************************************** //
+    // ********************* REQUIRED METHODS ******************** //
+    // *********************************************************** //
+
+    /// Some formats, such as `dtrace`, contain a header or other non-stack
+    /// information at the beginning of their input files. If header information
+    /// is present, this method **must** consume it (i.e. advance the provided
+    /// reader past it).
+    ///
+    /// This method also provides an opportunity to do processing of actual
+    /// stack data on the main thread before worker threads are spun up. For
+    /// an example of why this might be necessary, see `perf`, which can require
+    /// reading the first stack in order to know how to process the rest.
+    ///
+    /// If the format you are working with does not contain header information
+    /// or does not need any special, up-front processing, just have this method
+    /// return `Ok(())` immediately.
+    fn pre_process<R>(&mut self, reader: &mut R, occurrences: &mut Occurrences) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// The primary method.
```

Comment:
Same as above -- first line should be brief and descriptive. "The primary method" doesn't convey much information :)
```rust
/// Primary driving method for the trait to process all samples in a chunk of input.
/// 
```